### PR TITLE
Return Service Unavailable 503 when every backend is down

### LINKED ISSUE #436: Return 503 Service unavailable when all backend is unavailable
Currently when all backends are down we return 500 internal server error by default.
But I think we will have to return service unavailable.
500 internal error should be the errors in cases of problems on the Carapace side, such as:
- no action configured
- no backend configured
that is, errors internal to Carapace that are not attributable to problems on the backend side.

### PR DESCRIPTION
**Pull request by hamadodene**
Closes #436

### CODE REVIEW

**Review by NiccoMlt — CHANGES_REQUESTED** [review]
Looks good to me, but you need to fix a couple of tests:

- org.carapaceproxy.backends.StuckRequestsTest ► testBackendUnreachableOnStuckRequest(true)
- org.carapaceproxy.server.mapper.BasicStandardEndpointMapperTest ► testRouteErrors

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 1]
```diff
@@ -276,13 +277,22 @@ public SimpleHTTPResponse mapInternalError(String routeid) {
...
         }
+        // custom global
+        if (errorAction == null && defaultServiceUnavailable != null) {
+            errorAction = actions.get(defaultServiceUnavailable);
+        }
+        if (errorAction != null) {
+            return new SimpleHTTPResponse(errorAction.getErrorCode(), errorAction.getFile(), errorAction.getCustomHeaders());
+        }
         // fallback
         return super.mapServiceUnavailableError(routeId);
```
Comment (NiccoMlt) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java:297]: We can leverage the `Optional` better here, i.e. like it follows:

```java
// custom for route
Optional<RouteConfiguration> config = routes.stream().filter(r -> r.getId().equalsIgnoreCase(routeid)).findFirst();
ActionConfiguration errorAction = config
        .map(RouteConfiguration::getErrorAction)
        .map(actions::get)
        // custom global
        .orElseGet(() -> defaultInternalErrorAction != null ? actions.get(defaultInternalErrorAction) : null);
if (errorAction != null) {
    return new SimpleHTTPResponse(errorAction.getErrorCode(), errorAction.getFile(), errorAction.getCustomHeaders());
}
// fallback
return super.mapInternalError(routeid);
```

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 2]
```diff
@@ -276,13 +277,22 @@ public SimpleHTTPResponse mapInternalError(String routeid) {
...
+                errorAction = actions.get(action);
             }
         }
+        // custom global
+        if (errorAction == null && defaultServiceUnavailable != null) {
+            errorAction = actions.get(defaultServiceUnavailable);
+        }
+        if (errorAction != null) {
+            return new SimpleHTTPResponse(errorAction.getErrorCode(), errorAction.getFile(), errorAction.getCustomHeaders());
+        }
```
Comment (NiccoMlt) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java:295]: This part is pretty much identical in:

- `mapInternalError`: lines 258-273
- `mapServiceUnavailableError`: lines 280-295

Maybe we should group it in an external method?

**Comment by hamadodene** [comment]
@NiccoMlt test Fixed

**Review by NiccoMlt — CHANGES_REQUESTED** [review]
Good job! Just a nitpick and we can go

**Review by NiccoMlt — APPROVED** [review]
Good job!

**Pull request merged**